app/src/main/java/org/mozilla/focus/fragment/BrowserFragment.kt

```diff
@@ -44,6 +44,12 @@ class BrowserFragment : WebFragment(), CursorEvent {
         }
     }
 
+    // WebFragment expects a value for these properties before onViewCreated. We use a getter
+    // for the properties that reference session because it is lateinit.
+    override lateinit var session: Session
```

Comment:
When is the session set? I remember looking through the Session code and seeing it was pretty complicated - is SessionManager in charge of returning the saved session or creating a new session?

Reply:
In onCreate, I called `initSession`:

```kotlin
        val sessionUUID = arguments.getString(ARGUMENT_SESSION_UUID)
                ?: throw IllegalAccessError("No session exists")
        session = if (sessionManager.hasSessionWithUUID(sessionUUID))
            sessionManager.getSessionByUUID(sessionUUID)
        else
            NullSession()

        setBlockingEnabled(session.isBlockingEnabled)
        session.url.observe(this, Observer { url -> this@BrowserFragment.url = url })
        session.loading.observe(this, object : NonNullObserver<Boolean>() {
            public override fun onValueChanged(loading: Boolean) {
                val activity = activity as MainActivity
                updateCursorState()
                if (!loading && activity.isReloadingForYoutubeDrawerClosed) {
                    activity.isReloadingForYoutubeDrawerClosed = false

                    // We send a play event which:
                    // - If we're on the video selection page, does nothing.
                    // - If we're in a fullscreen video, will show the play/pause controls on the screen so
                    // we don't just see a black screen.
                    activity.dispatchKeyEvent(KeyEvent(KeyEvent.ACTION_UP, KeyEvent.KEYCODE_MEDIA_PLAY_PAUSE))
                }
            }
        })
```

so SessionManager returns an instance in `onCreate` – it could be immutable if fragments didn't have a lifecycle.